libbpf-tools/memleak.bpf.c

```diff
@@ -0,0 +1,392 @@
+
+	const struct alloc_info *info = bpf_map_lookup_elem(&allocs, &addr);
+	if (!info)
+		return 0;
+
+	bpf_map_delete_elem(&allocs, &addr);
+	update_statistics_del(info->stack_id, info->size);
+
+	if (trace_all) {
+		bpf_printk("free entered, address = %lx, size = %lu\\n",
+				address, info->size);
+	}
+
+	return 0;
+}
+
+SEC("uprobe")
+int BPF_KPROBE(malloc_enter, size_t size)
+{
+	return gen_alloc_enter(size);
+}
+
+SEC("uretprobe")
+int BPF_KRETPROBE(malloc_exit)
+{
+	return gen_alloc_exit(ctx);
+}
+
+SEC("uprobe")
+int BPF_KPROBE(free_enter, void *address)
+{
+	return gen_free_enter(address);
+}
+
+SEC("uprobe")
+int BPF_KPROBE(calloc_enter, size_t nmemb, size_t size)
+{
+	return gen_alloc_enter(nmemb * size);
+}
+
+SEC("uretprobe")
+int BPF_KRETPROBE(calloc_exit)
+{
+	return gen_alloc_exit(ctx);
+}
+
+SEC("uprobe")
+int BPF_KPROBE(realloc_enter, void *ptr, size_t size)
+{
+	gen_free_enter(ptr);
+
+	return gen_alloc_enter(size);
+}
+
+SEC("uretprobe")
+int BPF_KRETPROBE(realloc_exit)
+{
+	return gen_alloc_exit(ctx);
+}
+
+SEC("uprobe")
+int BPF_KPROBE(mmap_enter)
+{
+	size_t size = (size_t)PT_REGS_PARM2(ctx);
+
+	return gen_alloc_enter(size);
+}
+
+SEC("uretprobe")
+int BPF_KRETPROBE(mmap_exit)
+{
+	return gen_alloc_exit(ctx);
+}
+
+SEC("uprobe")
+int BPF_KPROBE(munmap_enter, void *address) {
+	return gen_free_enter(address);
+}
+
+SEC("uprobe")
+int BPF_KPROBE(posix_memalign_enter, void **memptr, size_t alignment, size_t size)
+{
+	const u64 memptr64 = (u64)(size_t)memptr;
+	const u64 pid = bpf_get_current_pid_tgid() >> 32;
+	bpf_map_update_elem(&memptrs, &pid, &memptr64, BPF_ANY);
+
+	return gen_alloc_enter(size);
+}
+
+SEC("uretprobe")
+int BPF_KRETPROBE(posix_memalign_exit)
+{
+	const u64 pid = bpf_get_current_pid_tgid() >> 32;
+	u64 *memptr64;
+	void *addr;
+
+	memptr64 = bpf_map_lookup_elem(&memptrs, &pid);
+
+	if (!memptr64)
+		return 0;
+
+	bpf_map_delete_elem(&memptrs, &pid);
+
+	if (bpf_probe_read_user(&addr, sizeof(void*), (void*)(size_t)*memptr64))
+		return 0;
+
+	const u64 addr64 = (u64)(size_t)addr;
+
+	return gen_alloc_exit2(ctx, addr64);
+}
+
+SEC("uprobe")
+int BPF_KPROBE(aligned_alloc_enter, size_t alignment, size_t size)
+{
+	return gen_alloc_enter(size);
+}
+
+SEC("uretprobe")
+int BPF_KRETPROBE(aligned_alloc_exit)
+{
+	return gen_alloc_exit(ctx);
+}
+
+SEC("uprobe")
+int BPF_KPROBE(valloc_enter, size_t size)
+{
+	return gen_alloc_enter(size);
+}
+
+SEC("uretprobe")
+int BPF_KRETPROBE(valloc_exit)
+{
+	return gen_alloc_exit(ctx);
+}
+
+SEC("uprobe")
+int BPF_KPROBE(memalign_enter, size_t alignment, size_t size)
+{
+	return gen_alloc_enter(size);
+}
+
+SEC("uretprobe")
+int BPF_KRETPROBE(memalign_exit)
+{
+	return gen_alloc_exit(ctx);
+}
+
+SEC("uprobe")
+int BPF_KPROBE(pvalloc_enter, size_t size)
+{
+	return gen_alloc_enter(size);
+}
+
+SEC("uretprobe")
+int BPF_KRETPROBE(pvalloc_exit)
+{
+	return gen_alloc_exit(ctx);
+}
+
+SEC("tracepoint/kmem/kmalloc")
+int tracepoint__kmalloc(struct trace_event_raw_kmem_alloc *ctx)
```

Comment:
this `tracepoint__` prefix is BCC necessity, but for libbpf is just a waste, let's drop this prefix altogether. It would be better to have memleak-specific short prefix to distinguish it from other programs and maps in the system. E.g. "mlk_" or even "memleak_" would be helpful